File: spark-scala/spark-datasources/spark-jdbc/spark-jdbc-sqlserver/src/main/scala/com/mahfooz/spark/jdbc/read/SparkSqlJdbc.scala
package com.mahfooz.spark.jdbc.read

import com.mahfooz.spark.jdbc.config.ConfigUtil
import org.apache.spark.sql.SparkSession

object SparkSqlJdbc {

  def main(args: Array[String]): Unit = {

    val spark = SparkSession
      .builder()
      .master("local[*]")
      .appName(getClass.getSimpleName)
      .getOrCreate()

    val tableName = "master.sys.all_views"

    val jdbcProperties = ConfigUtil.getOdbcProperty("db.conf")
    val jdbcDF = spark.read
      .format("jdbc")
      .option(
        "url",
        jdbcProperties.getProperty("url")
      )
      .option("dbtable", tableName)
      .option("user", jdbcProperties.getProperty("user"))
      .option("password", jdbcProperties.getProperty("password"))
      .load()

    jdbcDF.show(false)
  }
}

File: spark-java/spark-dataframe/src/main/java/com/mahfooz/spark/dataframe/ds/SparkSqlJdbc.java
package com.mahfooz.spark.dataframe.ds;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

import java.util.Properties;

public class SparkSqlJdbc {

    public static void main(String[] args) {
        SparkSession spark = SparkSession
                .builder()
                .appName("Java Spark SQL data sources example")
                .config("spark.some.config.option", "some-value")
                .getOrCreate();

        runJdbcDatasetExample(spark);
    }
    private static void runJdbcDatasetExample(SparkSession spark) {

        // $example on:jdbc_dataset$
        // Note: JDBC loading and saving can be achieved via either the load/save or jdbc methods
        // Loading data from a JDBC source
        Dataset<Row> jdbcDF = spark.read()
                .format("jdbc")
                .option("url", "jdbc:postgresql:dbserver")
                .option("dbtable", "schema.tablename")
                .option("user", "username")
                .option("password", "password")
                .load();

        Properties connectionProperties = new Properties();
        connectionProperties.put("user", "username");
        connectionProperties.put("password", "password");
        Dataset<Row> jdbcDF2 = spark.read()
                .jdbc("jdbc:postgresql:dbserver", "schema.tablename", connectionProperties);

        // Saving data to a JDBC source
        jdbcDF.write()
                .format("jdbc")
                .option("url", "jdbc:postgresql:dbserver")
                .option("dbtable", "schema.tablename")
                .option("user", "username")
                .option("password", "password")
                .save();

        jdbcDF2.write()
                .jdbc("jdbc:postgresql:dbserver", "schema.tablename", connectionProperties);

        // Specifying create table column data types on write
        jdbcDF.write()
                .option("createTableColumnTypes", "name CHAR(64), comments VARCHAR(1024)")
                .jdbc("jdbc:postgresql:dbserver", "schema.tablename", connectionProperties);
        // $example off:jdbc_dataset$
    }
}
